roms_tools/setup/cdr_release.py

```diff
@@ -0,0 +1,481 @@
+        - If no times are provided, assumes a constant concentration throughout.
+
+        Parameters
+        ----------
+        times : list
+            List of datetime-like time points.
+        start_time : datetime-like
+            Start of the interval to cover.
+        end_time : datetime-like
+            End of the interval to cover.
+
+        Returns
+        -------
+        self : Concentration
+            Updated instance with extended values.
+        """
+        start_pad = self.values[0] if isinstance(self.values, list) else self.values
+        end_pad = self.values[-1] if isinstance(self.values, list) else self.values
+        return self._extend_scalar_series(
+            times, start_time, end_time, start_pad=start_pad, end_pad=end_pad
+        )
+
+
+class Release(BaseModel):
+    """Defines the basic properties and timing of a carbon dioxide removal (CDR)
+    release.
+
+    Attributes
+    ----------
+    name : str
+        Unique identifier for the release.
+    lat : float
+        Latitude of the release location in degrees North. Must be between -90 and 90.
+    lon : float
+        Longitude of the release location in degrees East.
+    depth : float
+        Depth of the release in meters. Must be non-negative.
+    hsc : float
+        Horizontal scale (standard deviation) of the release in meters. Must be non-negative.
+    vsc : float
+        Vertical scale (standard deviation) of the release in meters. Must be non-negative.
+    times : list of datetime
+        Time points of the release events. Must be strictly increasing and within the simulation window.
+    """
+
+    name: str
+    lat: Annotated[float, Ge(-90), Le(90)]
+    lon: float
+    depth: NonNegativeFloat
+    hsc: NonNegativeFloat = 0.0
+    vsc: NonNegativeFloat = 0.0
+    times: List[datetime]
+
+    model_config = ConfigDict(extra="forbid")
+
+    @model_validator(mode="after")
+    def check_increasing_times(self) -> "Release":
+        """Validates that `times` are strictly increasing and fall within the specified
+        time window.
+
+        Raises
+        ------
+        ValueError
+            If times are not strictly increasing, or fall outside the [start_time, end_time] window.
+        """
+
+        if self.times and len(self.times) > 0:
+            if not all(t1 < t2 for t1, t2 in zip(self.times, self.times[1:])):
+                raise ValueError(
+                    f"'times' must be strictly monotonically increasing. Got: {self.times}"
+                )
+        return self
+
+    def extend_times_to_endpoints(self, start_time, end_time) -> None:
+        """Ensures that `times` includes both `start_time` and `end_time`.
+
+        Modifies `self.times` in place by prepending or appending times as needed.
+        If `times` is empty, it will be set to [`start_time`, `end_time`].
+        """
+
+        if not self.times:
+            self.times = [start_time, end_time]
```

Comment:
It seems like this base case should not be part of this method. Initializing `self.times` is not the same as extending, so we're doing 2 things in one place. I have a feeling that you really want something in a `post_init` method to initialize `self.times`

Reply:
We cannot initialize `self.times` with `[start_time, end_time]` in `post_init` because subsequent checks verify that the number of times matches the number of provided fluxes and concentrations. Initializing early would cause these checks to fail due to a mismatch.